Add GDT header

chrapati24 opened the pull request:
Just the latest PR but in draft.

chrapati24 committed:
Add GDT header. Waiting checking before continue.

keyboard-slayer commented on src/kernel/arch/x86_64/gdt/gdt.h:
```diff
@@ -0,0 +1,23 @@
+
+struct gdt_entry
+{
+    u16int base_low;
```
The types you are using in your struct are not defined at all.
Instead you should use the type that are defined in  `#include <stdint.h>`

keyboard-slayer commented on src/kernel/arch/x86_64/gdt/gdt.h:
```diff
@@ -0,0 +1,23 @@
+struct gdt_ptr
+{
+    u16int limit;
+    u32bit base;
```
Instead of 32 bits int for the base, you should use a `uintptr_t`. This type represent the address of a pointer

keyboard-slayer commented on src/kernel/arch/x86_64/gdt/gdt.h:
```diff
@@ -0,0 +1,23 @@
+
+// Entry of table
+
+struct gdt_entry
```
You should do a typedef for this struct.

Also this GDT doesn't seems right, can you double check the documentation from [osdev](https://wiki.osdev.org/GDT_Tutorial)

d0p1s4m4 replied:
I would also recommend reading of [Intel manual Vol. 3A](https://www.intel.com/content/dam/support/us/en/documents/processors/pentium4/sb/25366821.pdf) which explain a lot about GDT.

keyboard-slayer commented on src/kernel/arch/x86_64/gdt/gdt.h:
```diff
@@ -0,0 +1,23 @@
+
+#pragma once
+
+// Entry of table
```
You should remove the commentary. 
The code is self explanatory.

keyboard-slayer commented on src/kernel/arch/x86_64/gdt/gdt.h:
```diff
@@ -0,0 +1,23 @@
+    u8int access_byte;
+    u8int base_middle;
+    u8int base_high;
+}__attribute__((packed));
```
You should define a macro in [here](https://github.com/keyboard-slayer/cpcdos-os3/blob/main/src/libs/macro.h), and also include that header.

This make the code cleaner :wink:

keyboard-slayer commented on src/kernel/arch/x86_64/gdt/gdt.h:
```diff
@@ -0,0 +1,23 @@
+    u16int limit;
+    u32bit base;
+}__attribute__((packed));
+
```
Side note: you don't need to create a new folder for the GDT.

Just put the header in the `src/kernel/arch/x86_64/` folder

keyboard-slayer commented:
I also didn't see the struct for the TSS in your code. If you want more information about it, you can find some explanations here [here](https://wiki.osdev.org/Global_Descriptor_Table#Long_Mode_System_Segment_Descriptor)

keyboard-slayer commented:
Also you didn't respect the commit style (check the link on the readme of the repo)

chrapati24 committed:
fix(gdt):changes about previous review

keyboard-slayer commented on src/kernel/arch/x86_64/gdt.c:
```diff
@@ -0,0 +1,6 @@
+#include "gdt.h"
```
You should include the local headers after the global headers

keyboard-slayer commented on src/kernel/arch/x86_64/gdt.h:
```diff
@@ -12,12 +12,14 @@ struct gdt_entry
 }__attribute__((packed));
-
-// Pointer of table
+typedef struct gdt_entry gdt_entry_t;
```
You should do the typedef like this

```C
typedef struct
{
    ...
} GdtEntry;
```
Typedefs should be in a class like notation

keyboard-slayer commented on src/kernel/arch/x86_64/gdt.h:
```diff
@@ -12,12 +12,14 @@ struct gdt_entry
     u8int base_middle;
     u8int base_high;
 }__attribute__((packed));
```
You didn't add `__attribute__((packed))` in `macro.h`

This will allow us to do something like

```c
typedef struct packed
{
   ...
};
```

keyboard-slayer commented on src/kernel/arch/x86_64/gdt.h:
```diff
@@ -12,12 +12,14 @@ struct gdt_entry
 
 struct gdt_ptr
 {
     u16int limit;
```
You didn't modify the types as requested previously

keyboard-slayer commented on src/kernel/arch/x86_64/gdt.h:
```diff
@@ -12,12 +12,14 @@ struct gdt_entry
 }__attribute__((packed));
+typedef struct gdt_ptr gdt_ptr_t;
+
+void gdt_init();
```
In C, when you don't give any argument to the function you add a `void` in the argument

like this:

```C
void gdt_init(void);
```

chrapati24 committed:
Merge branch 'keyboard-slayer:main' into main

chrapati24 committed:
add(gdt) : adding gdt.c file

chrapati24 committed:
Merge branch 'main' of https://github.com/chrapati24/cpcdos-os3

chrapati24 committed:
fix(macro.h) : add packed attribute

fix(gdt.h) : Refactoring of code

add : gdt.c & gdt.S for initialization of Global Desc. Table

d0p1s4m4 commented on src/kernel/arch/x86_64/gdt.h:
```diff
@@ -4,22 +4,28 @@
-typedef struct gdt_entry gdt_entry_t;
-
-struct gdt_ptr
+    uint16_t *limitLow;
+    uint16_t *baseLow;
+    uint8_t *baseMiddle;
+    uint8_t *baseHigh;
+    uint8_t *granularity;
+    uint8_t *access;
+    uint8_t *flags;
+
```
why ptr ?

d0p1s4m4 commented on src/kernel/arch/x86_64/gdt.h:
```diff
@@ -4,22 +4,28 @@
-    uintptr_t base;
-}__attribute__((packed));
-typedef struct gdt_ptr gdt_ptr_t;
+    uintptr_t *limit;
+    uint16_t *base;
```
why ptr ?

chrapati24 replied:
I can change with other type if you want. With uint32_t.

d0p1s4m4 commented on src/kernel/arch/x86_64/gdt.h:
```diff
@@ -4,22 +4,28 @@
-typedef struct gdt_ptr gdt_ptr_t;
+    uintptr_t *limit;
+    uint16_t *base;
+}GDTPtr;
```
add space between } and GDTPtr;

keyboard-slayer commented:
Closing pr since [dd6b688](https://github.com/keyboard-slayer/cpcdos-os3/commit/dd6b688d1ae352a77ec5d4f64a5dd648b3af6cd8)
